### ISSUE TITLE
Splits invalid when collection order not deterministic

### ISSUE THREAD

**Issue by mbkroese** [jerry-git/pytest-split#25]
The big assumption underlying the two splitting algorithms is that the order of collected items is constant.
However, I've come across a case where this assumption was violated.
In my case I had a test parametrised with `pytest.mark.parametrize`, but the items to parametrize with would sometimes change order.

Take this example:

```python
import pytest

@pytest.mark.parametrize('name', set(['henk', 'ingrid']))
def test_hello(name):
    pass
```

If you run this often enough you'll see that the order changes:

```
[2021-06-17 22:47:10] test_temp.py::test_hello[henk] PASSED                                                                                                                [ 50%]
[2021-06-17 22:47:10] test_temp.py::test_hello[ingrid] PASSED                                                                                                              [100%]
```

and

```
[2021-06-17 22:47:10] test_temp.py::test_hello[ingrid] PASSED                                                                                                              [ 50%]
[2021-06-17 22:47:10] test_temp.py::test_hello[henk] PASSED                                                                                                                [100%]
```

I'm not sure if the `duration_based_chunks` algorithm can really avoid such cases. In the case of `least_duration` I can simply order the nodes alphabetically. This would break the "relative order" property, but I don't mind that and I don't expect this option to have any users yet considering we merged it so recently.

So concretely my proposal would be:
1. Warn users about this case for `duration_based_chunks`, but don't change anything.
2. Update the `least_duration` algorithm to be resilient to this problem.

**Comment by jerry-git** [comment]
I assume it's not deterministic here because of `set`, or can you repro it also with `list` or `tuple`?

**Comment by mbkroese** [comment]
No, this problem occurs when either the data structure has non-deterministic order or the code generating the parametrised test cases is for some reason not deterministic.

**Comment by jerry-git** [comment]
I think we could go with `1.` aka make sure the tests inside same `parametrize` are run in the same group. However, the downside is ofc that if one parametrised test is very time consuming vs the rest of the suite, the splits would not be great.

OTOH, maybe it's better to make sure that we don't accidentally skip tests (or run some test in multiple groups) 🤔 

With these thoughts, I'd go with `1.` 🙂

**Comment by mbkroese** [comment]
> downside is ofc that if one parametrised test is very time consuming vs the rest of the suite, the splits would not be great.

Yes, and I wonder if we should perhaps be safe by default (i.e. option 1) and allow users to do the unsafe thing (existing behaviour). If we make clear what the tradeoffs are, the user can then decide for him/herself. In other words add a parameter that `--split-level=func` by default but can be set to `--split-level=parametrized` or `--split-level=file`?

**Comment by dpanici** [comment]
Was this implemented yet? If not I think it would be a great feature to have, as I have some tests I want ran in the same group under the same parametrize decorator

**Comment by vokracko** [comment]
Hello, I've opened a chain of PRs to address the non-deterministic collection universally (superseding the fix for `least_duration` done in https://github.com/jerry-git/pytest-split/pull/52)
- #120
- #121
- #122
- #123
- #124

Could you have a look @jerry-git? 🙏
